src/swarm-lang/Swarm/Language/Load.hs

```diff
@@ -39,17 +44,61 @@ import Swarm.Language.Syntax.Import hiding (ImportPhase (..))
 import Swarm.Language.Syntax.Import qualified as Import
 import Swarm.Pretty (prettyText)
 import Swarm.Util (Encoding (SystemLocale), getModificationTimeMay, readFileMayT, showT)
+import Swarm.Util.GlobalCache qualified as GC
 import Swarm.Util.Graph (findCycleImplicit)
-import Swarm.Util.InternCache qualified as IC
 import Witch (into)
 
+-- ~~~~ Note [Module loading]
+--
+-- Loading modules is a bit tricky, since we have to deal with:
+--   - Resolving imports, especially relative imports, to a specific location
+--   - Loading imported modules from disk/network
+--   - Transitively loading imports, imports of imports, etc.
+--   - Checking for import cycles (+ not getting stuck in infinite recursion if so)
+--   - Making sure we don't reload the same module more than once
+--
+-- The goal is to read things in from disk and end up with a
+-- consistent set of modules, where every loaded module's imports have
+-- also been loaded, all modules are keyed by a fully resolved and
+-- canonicalized import location, and every import directive contained
+-- in source has also been resolved and canonicalized.  Note that
+-- typechecking and elaboration are done as separate phases
+-- after loading + resolution completes.
+--
+-- The top-level function is 'resolve', which calls 'resolveImports'
+-- and then 'checkImportCycles'.
+--
+-- 'resolveImports', 'resolveImport', and 'importModule' are mutually
+-- recursive and do the bulk of the work.  'resolveImports' traverses
+-- an AST and calls 'resolveImport' on each import it contains.
+-- 'resolveImport' processes a single import; it first checks whether
+-- the import has already been loaded, and if not, it calls
+-- 'importModule', which actually loads the module (via 'readLoc'),
+-- calls 'resolveImports' on it, and ensures (via 'validateImport')
+-- that it contains only @def@s.
+
 -- | A SourceMap associates canonical 'ImportLocation's to modules.
-type SourceMap phase = Map (ImportLoc (ImportPhaseFor phase)) (Module phase)
+--   It is an /ordered/ map, and has as an invariant that the modules
+--   are topologically ordered, i.e. if B imports A, then A comes
+--   before B in order.  This ensures that if we process modules in
+--   order, by the time we process a module we have already processed
+--   any modules it imports.
+type SourceMap phase = OMap (ImportLoc (ImportPhaseFor phase)) (Module phase)
+
+type ResLoc = ImportLoc Import.Resolved
```

Comment:
This type synonym is not exported, but appears in exported function signature - won't that look weird in Haddock? Maybe rename the synonym as `ResolvedImport` or `LoadedImport` with a short explanation.